scripts/generate-ffdhe.py

```diff
@@ -0,0 +1,514 @@
+	content = bytearray()
+	for entry in sequence:
+		content += entry
+	out += derLength(len(content))
+	out += content
+	return out
+
+def derUnsignedInteger(hexString):
+	'''
+		Encode a hex string to a ASN.1 DER INTEGER.
+		This function only handles unsigned integers.
+
+		Returns a bytearray of the resulting data.
+
+		The encoding of an ASN.1 DER INTEGER is an
+		identifier (0x02), a length (the length of the
+		content bytes -- see the derLength function),
+		and the content, which in this case is the
+		bytes that make up the integer.
+
+		An INTEGER can be signed, or unsigned. Our
+		function here only deals with unsigned integers.
+
+		Signed integers have the MSB/sign bit set.
+
+		Our input data is guaranteed to be unsigned,
+		so if we detect that the MSB is set in the
+		data, we prepend a zero byte to it. This
+		is the way to signal that the value is unsigned
+		rather than signed.
+	'''
+	out = bytearray((0x02,)) # INTEGER tag
+	buf = dehexify(hexString)
+
+	# If the sign bit is set, prepend a zero byte.
+	# Otherwise, the number will be treated as signed,
+	# which we don't want.
+	if (buf[0] & 0x80) != 0:
+		buf = bytearray((0,)) + buf
+	out += derLength(len(buf))
+	out += buf
+	return out
+
+def pem(der, kind='DH PARAMETERS'):
+	'''
+		Convert the ASN.1 DER data to PEM form.
+
+		Returns the resulting string (type 'unicode' for Python 2).
+
+		The PEM format consists of a header, the content itself (base64,
+		using the standard alphabet) and a trailer.
+
+		The header is of the form:
+			-----BEGIN $KIND-----
+
+		and the trailer is of the form:
+			-----END $KIND-----
+
+		where $KIND defines the type of data
+		contained in the PEM data.
+
+		Between the header and the trailer is
+		the content itself. The content is the
+		DER form encoded using base64.
+
+		Our implementation splits the base64
+		data such that each line of base64 data
+		can be no longer than 64 characters long.
+	'''
+	out = u'-----BEGIN {0}-----\n'.format(kind)
+	enc = base64.b64encode(der)
+	n = 0
+	for ch in enc:
+		if type(ch) == int: # Python 3
+			out += chr(ch)
+		else:
+			out += ch
+		n += 1
+		if n == 64:
+			out += '\n'
+			n = 0
+	if out[-1] != '\n':
+		out += '\n'
+	out += u'-----END {0}-----\n'.format(kind)
+	return out
+
+def writeHeader(f):
+	'''
+		Output the LICENSE.header file plus a warning
+		about the output file being auto-generated.
+
+	'''
+	hf = codecs.open('LICENSE.header', 'r', 'utf-8')
+	hdr = hf.read()
+	hf.close()
+
+	f.write(hdr)
+	f.write('\n')
+	f.write('// Automatically generated by "generate-ffdhe.py". DO NOT EDIT BY HAND.\n')
+	f.write('//\n')
+	f.write('// I also agree to have manually vetted this file for correctness.\n')
+	f.write('//\n')
+	f.write('// If I do not agree, I will not have removed the line above saying\n')
+	f.write('// otherwise. Nor will I have removed the line below this one which\n')
+	f.write('// will cause a preprocessor error. Oops!\n')
+	f.write('#error Please verify this file is correct\n')
```

Comment:
Do we generate the file and then the committer has to verify the output and have to modify the auto-generated files comments??
Should be documented. And I don't really like/understand the two-split (non-)agreement. Code comments should not talk about the author as *I*.